fix(linux): use PulseAudio recorder (parecord) instead of fmedia as library

saif-ellafi opened the pull request:
Hey all, looking to help with: https://github.com/llfbandit/record/issues/461

This is to use PulseAudio  `parecord` tool which is typically widely available in Linux Distributions, instead of the deprecated and abandoned fmedia project originally used in this library.

### Needs testing help!!

**Important:**
1. Seems to be working fine, tested with m4a, start, stop, pause, and resume.
2. Pause and Resume kills the process, and then starts another process under the same pipe. Needs testing, or if too risky, disable the support for pause and resume. It seems to work though.
3. Have not tested all codecs, testing help needed. Tested on Fedora Workstation.
4. Listing devices works, and seems to be working fine. Using `pactl` which is a standard library.
5. I am not sure if we still really need to have gstreamer1 and gstreamer1-plugins-base installed for this work, I have not removed those dependencies, if anyone can test, appreciated, but currently using PulseAudio is great for packaging into Flatpak or other standard distribution frameworks.
6. Only 1 and 2 channels are supported, but if you see the opportunity to remap channels and test it, maybe it works?

Best regards
Saif

saif-ellafi commented:
alsa-utils for arecord may be a pain to add to flatpak builds of the flutter apps... will continue testing and figuring out what is the most reliable.

saif-ellafi commented:
I moved to PulseAudio instead of ALSA which is a more modern layer and easier to package into standards (i.e. Flatpak).
However, for some reason, `parecord` needs 2 seconds of recording time to not cut off the recording at the end, extremely weird, but otherwise seems to work fine.

llfbandit commented:
Many thanks for your contribution!

llfbandit commented on record_linux/lib/record_linux.dart:
```diff
@@ -65,76 +66,82 @@ class RecordLinux extends RecordPlatform {
-    await _callFMedia(['--globcmd=stop'], recorderId: recorderId);
-    await _callFMedia(['--globcmd=quit'], recorderId: recorderId);
+    _parecordProcess?.kill();
+    _parecordProcess = null;
```
How ffmpeg process is stopped?
How to ensure that ffmpeg process has finished before sending state update?

saif-ellafi replied:
since ffmpeg is being piped, it will stop on its own as soon as it finishes the processing of its buffer. Force terminating it earlier can be risky. I will however set it back to null which I missed!

llfbandit commented on record_linux/lib/record_linux.dart:
```diff
@@ -1,13 +1,12 @@
 import 'dart:async';
 import 'dart:convert';
 import 'dart:io';
+import 'dart:ui';
```
Is dart:ui import necessary?

saif-ellafi replied:
This is for VoidCallback. Will change to import 'package:flutter/foundation.dart'; which is more typical.

llfbandit commented on record_linux/lib/record_linux.dart:
```diff
@@ -251,90 +191,83 @@ class RecordLinux extends RecordPlatform {
-    String recorderId,
-    List<String> out,
-  ) async {
+  List<InputDevice> _parseInputDevices(List<String> output) {
```
Add output sample of listing to better understand the parsing

saif-ellafi replied:
Sure thing, by the way currently the id being stored is the Source number which changes on reboot or system logouts. Must look into the dump of this command to see if there is a more permanent device id!

llfbandit commented on record_linux/lib/record_linux.dart:
```diff
@@ -251,90 +191,83 @@ class RecordLinux extends RecordPlatform {
+      case AudioEncoder.aacHe:
+        return ['-c:a', 'aac', '-b:a', '${bitRate/1000}k', path];
+      case AudioEncoder.wav:
+        return ['-c:a', 'pcm_s16le', path];
```
I guess this may lead to issues here. As far as I understand if file extension is wav this will provide WAVE format?

saif-ellafi replied:
I have improved this in a way that flac and wav are directly done by `parecord` which can handle them. The others work just fine with ffmpeg (I tried wav, flac, and aac). 

Not sure if you see any improvements room?

llfbandit reviewed (changes requested):
I'm blind but I tried :)

Few thoughts:
- We will need some updated docs (external dependencies)
- Can we implement stream by using parecord?
- Can we add other formats for this platform? (PCM, AMRx, ...)

saif-ellafi commented on record_linux/lib/record_linux.dart:
```diff
@@ -65,76 +66,82 @@ class RecordLinux extends RecordPlatform {
     if (_state == RecordState.record) {
-      await _callFMedia(['--globcmd=pause'], recorderId: recorderId);
-
+      _parecordProcess?.kill(ProcessSignal.sigstop);
```
I honestly do not know how to test multiple recorders, the class was holding a single process anyway, so pause and resume behave the same for single recording. I am open to ideas, or it can be improved in the future?

saif-ellafi commented on record_linux/lib/record_linux.dart:
```diff
@@ -65,76 +66,82 @@ class RecordLinux extends RecordPlatform {
-      recorderId: recorderId,
-    );
+  String numChannels;
+  if (config.numChannels == 1 || config.numChannels == 2) {
```
Will do!

saif-ellafi commented:
@llfbandit I applied most of your comments, also for wav and flac we need one less command (and therefore less risk), since `parecord` can directly encode into thise.

gyoussef55 commented:
> I'm blind but I tried :)
> 
> Few thoughts:
> 
>     * We will need some updated docs (external dependencies)
> 
>     * Can we implement stream by using parecord?
> 
>     * Can we add other formats for this platform? (PCM, AMRx, ...)

for streaming I used `parec` before by fixing [flutter_audio_capture](https://pub.dev/packages/flutter_audio_capture) so implementing stream is possible

saif-ellafi commented:
Excuse my ignorance, @gyoussef55 @llfbandit what is the overall benefit of streaming? versus the current record and pipe to encode?

llfbandit commented:
Some may need PCM stream only to detect sound or silence, server-side encoding, frequency to make instrument tuner for example and so on...
So encoded files aren't our single target. That said, this is not a goal for this PR.

saif-ellafi commented:
Hello @llfbandit 

Any chances of seeing this one merged? I just updated main branch to it, and I also changed the ID to use the alsa node.name which is more reliable than the previous entry ID (it changed every system reboot).

If it helps something, I don't have more than a handful of users (4?) but I've pushed this fork branch over there and no issues so far on Linux for m4a codec. Of course more testing is good but I think this is already an improvement over a non-functioning Linux version!

llfbandit commented:
Before merging, there are few missing pieces related to documentation in README.md:
- the change to parecord should be stated.
- the need of ffmpeg dependency also (and associated license?).
- an update to feature matrix if needed.
-
- Update version of record_linux to 2.0.0. (I will release a version 1 for fmedia before to enable fallback).
- Update record_linux changelog to state parecord & ffmpeg usage.

Are you able to continue the support of linux plaform after this PR? As you know already, I don't have linux host so I can't support directly this platform anymore.

Finally, thanks again to contribute to this package.

rohanmars commented:
Nice work @saif-ellafi.
A good usecase for the streaming API is https://github.com/k2-fsa/sherpa-onnx/tree/master/flutter-examples/streaming_asr, a sample app for the sherpa-onnx voice to text project.
I implemented the startStreaming method on top of this PR with PCM support, it only required a few changes. All seems to work. More then happy to cleanup the code and submit it after this PR is merged. Then sherpa_onnx flutter will work on linux for real time voice reognition.

saif-ellafi commented:
Hello @llfbandit 
Sorry for the late response.

I cannot commit to continue to support, but as a user of the library and for my interest on it, gladly I can run tests and keep it alive if it breaks ^^ But I cannot fully cover all the endless possibilities of the Linux environment.

Been using this for +month and working fine so far and also for my (very few) group of Linux users.
- ffmpeg is Free Software GNU: https://www.ffmpeg.org/legal.html - should not be a problem at all
- ffmpeg as dependency is very very common, but also, it is only necessary for non standard formats. WAV for example can be directly encoded by pulse audio which is a huge advantage, and does not invoke ffmpeg to encode
- Linux developers should package their software approprietly, ffmpeg and pulse audio libs can be included in snaps and flatpaks for example

For changelog and metadata I ask you please you do that :)

Best regards
Saif

gyoussef55 commented:
Hey @llfbandit, is there a chance to add the startStreaming method that @rohanmars made?

rohanmars commented:
@gyoussef55 @llfbandit now that this PR was merged I'll clean up my code and submit a new PR with the streaming support.
